File: app/app/src/main/java/com/example/hakonsreader/api/responses/ListingResponse.java
package com.example.hakonsreader.api.responses;

import com.example.hakonsreader.api.model.RedditListing;
import com.google.gson.annotations.SerializedName;

import java.util.List;

/**
 * Response for any request that returns a list of listings
 */
public class ListingResponse  {

    @SerializedName("errors")
    private List<List<String>> errors;

    @SerializedName("data")
    private Data data;
    private static class Data {

        // For "more comments" kind of responses the array is called "things" instead of "children"
        @SerializedName(value = "children", alternate = "things")
        List<? extends RedditListing> listings;
    }

    /**
     * @return The listings retrieved from the response
     */
    public List<? extends RedditListing> getListings() {
        return data.listings;
    }

    /**
     * @return True if the response has errors
     */
    public boolean hasErrors() {
        return errors != null && !errors.isEmpty();
    }

    /**
     * Retrieves the list of errors. The inner list contains various error information. The first
     * object in the inner list contains an enum type that indicates what kind of error it is. The
     * error can be identified with {@link com.example.hakonsreader.api.enums.ResponseErrors}
     *
     * @return The list of errors for the request
     */
    public List<List<String>> getErrors() {
        return errors;
    }
}

File: app/app/src/main/java/com/example/hakonsreader/api/responses/ListingResponse.kt
package com.example.hakonsreader.api.responses

import com.example.hakonsreader.api.jsonadapters.ListingListAdapter
import com.google.gson.annotations.JsonAdapter
import com.google.gson.annotations.SerializedName

/**
 * Response for any request that returns a list of listings
 */
class ListingResponse<T> {

    @SerializedName("errors")
    private var errors: List<List<String>>? = null

    @SerializedName("data")
    private var data: Data? = null

    inner class Data {
        // For "more comments" kind of responses the array is called "things" instead of "children"
        @SerializedName(value = "children", alternate = ["things"])
        @JsonAdapter(ListingListAdapter::class)
        var listings: List<T>? = null

        @SerializedName("after")
        var after: String? = null
    }

    fun getAfter(): String? {
        return data?.after
    }

    fun getListings() : List<T>? {
        return data?.listings
    }

    fun hasErrors() : Boolean {
        return !errors.isNullOrEmpty()
    }

    fun getErrors() : List<List<String>>? {
        return errors
    }
}